tests/telegram/test_engine.py

```diff
@@ -0,0 +1,100 @@
+from unittest import mock
+
+import pytest
+
+from bottery.exceptions import ImproperlyConfigured
+from bottery.message import Message
+from bottery.platform.telegram import TelegramEngine, TelegramUser
+
+
+def test_telegram_user_with_last_name():
+    sender = {
+        'id': 1,
+        'first_name': 'Andrew',
+        'last_name': 'Martin',
+    }
+
+    user = TelegramUser(sender)
+    assert str(user) == 'Andrew Martin (1)'
+
+
+def test_telegram_user_without_last_name():
+    sender = {
+        'id': 1,
+        'first_name': 'Andrew',
+    }
+
+    user = TelegramUser(sender)
+    assert str(user) == 'Andrew (1)'
+
+
+def test_engine_default_mode(token, session):
+    engine = TelegramEngine(token=token, session=session)
+    assert engine.mode == 'polling'
+
+
+def test_engine_set_mode(token, session):
+    engine = TelegramEngine(token=token, session=session, mode='webhook')
+    assert engine.mode == 'webhook'
+
+
+@pytest.mark.asyncio
+async def test_engine_missing_configuration_method(token, session):
+    engine = TelegramEngine(token=token, session=session,
+                            mode='not-defined-mode')
+
+    with pytest.raises(ImproperlyConfigured):
+        await engine.configure()
+
+
+@pytest.mark.asyncio
+async def test_engine_configure_polling(token, mock_session):
+    session = mock_session({'ok': True})
+    engine = TelegramEngine(token=token, session=session)
+
+    await engine.configure()
+
+    url = 'https://api.telegram.org/bot{}/deleteWebhook'.format(token)
+    engine.session.post.assert_called_once_with(url, json={})
+    assert engine.tasks == [engine.polling]
+
+
+@pytest.mark.skip(reason="Don't know why the mock below doesn't work :/")
+@pytest.mark.asyncio
+async def test_engine_configure_webhook(token, mock_session):
```

Comment:
@nicoddemus can you take a look at why this mock isn't working? :)

Reply:
Sure, will take a look tomorrow!

Reply:
Sorry for taking so long!

The problem is that you are calling `mock.patch` directly (from `unittest.mock`), instead of using it as a context manager:

```python
@pytest.mark.asyncio
async def test_engine_configure_webhook(token, mock_session):
    with mock.patch('bottery.platform.telegram.settings') as mocked_settings:
        mocked_settings.HOSTNAME = 'localhost'

        session = mock_session({'ok': True})
```

Because you are calling it directly, its `__enter__` method never gets called and then the actual mocking never happens.

You probably meant to use the `mocker` fixture:

```python
@pytest.mark.asyncio
async def test_engine_configure_webhook(token, mock_session, mocker):
    mocked_settings = mocker.patch('bottery.platform.telegram.settings')
    mocked_settings.HOSTNAME = 'localhost'

    session = mock_session({'ok': True}) 
```

`mocker` is the fixture provided by `pytest-mock` which allows you to use inside tests without having to write a ton of `with` statements. There's more details in the [rationale](https://github.com/pytest-dev/pytest-mock#why-bother-with-a-plugin) section of the README.

After changing to the `mocker` fixture, I get this exception:

```
tests\telegram\test_engine.py:62 (test_engine_configure_webhook)
token = 'telegram-api-token'
mock_session = <function mock_session.<locals>.session at 0x00000218713C7A60>
mocker = <pytest_mock.MockFixture object at 0x0000021873BC19B0>

    @pytest.mark.asyncio
    async def test_engine_configure_webhook(token, mock_session, mocker):
        mocked_settings = mocker.patch('bottery.platform.telegram.settings')
        mocked_settings.HOSTNAME = 'localhost'
    
        session = mock_session({'ok': True})
        engine = TelegramEngine(token=token, session=session, mode='webhook')
    
>       await engine.configure()

test_engine.py:71: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
..\..\bottery\platform\telegram.py:135: in configure
    await configure_mode()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <bottery.platform.telegram.TelegramEngine object at 0x0000021873C0BF28>

    async def configure_webhook(self):
        hostname = getattr(settings, 'HOSTNAME')
        if not hostname:
            raise ImproperlyConfigured('Missing HOSTNAME setting')
    
        response = await self.api.set_webhook(url=hostname)
        response = await response.json()
        if response['ok']:
            logger.debug('[%s] Webhook mode set', self.engine_name)
    
>       self.server.router.add_post('/', self.webhook)
E       AttributeError: 'TelegramEngine' object has no attribute 'server'
```

😁

Reply:
Thanks @nicoddemus!! Sorry for the delay too :/ Yep, it's working know. I tried the `@mock.path` decorator, but didn't work too, which is strange. But anyway :)